internal/net/buffer.go

```diff
@@ -0,0 +1,203 @@
+// SPDX-FileCopyrightText: 2023 The Pion community <https://pion.ly>
+// SPDX-License-Identifier: MIT
+
+package net
+
+import (
+	"errors"
+	"io"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/pion/transport/deadline"
+)
+
+// AddrPacket is a packet payload and the associated remote address from which
+// it was received. It is meant to be re-usable, so the length of the data bytes
+// may not reflect the length of the payload, while the size should.
+type AddrPacket struct {
+	addr net.Addr
+	data []byte
+	size int
+}
+
+// PacketBuffer is a circular buffer for network packets. Each slot in the
+// buffer supports a
+type PacketBuffer struct {
+	mutex sync.Mutex
+
+	packets     []AddrPacket
+	write, read int
+
+	// full indicates whether the buffer is full, which is needed to distinguish
+	// when the write pointer and read pointer are at the same index.
+	full bool
+
+	notify chan struct{}
+	closed bool
+
+	readDeadline *deadline.Deadline
+}
+
+// NewPacketBuffer creates a new PacketBuffer.
+func NewPacketBuffer() *PacketBuffer {
+	return &PacketBuffer{
+		readDeadline: deadline.New(),
+		// In the narrow context in which this package is currently used, there
+		// will always be at least one packet written to the buffer. Therefore,
+		// we opt to allocate with size of 1 during construction, rather than
+		// waiting until that first packet is written.
+		packets: make([]AddrPacket, 1),
+		full:    false,
+	}
+}
+
+// WriteTo writes a single packet to the buffer. The supplied address will
+// remain associated with the packet.
+func (b *PacketBuffer) WriteTo(p []byte, addr net.Addr) (int, error) {
+	b.mutex.Lock()
+
+	if b.closed {
+		b.mutex.Unlock()
+		return 0, io.ErrClosedPipe
+	}
+
+	var notify chan struct{}
+	if b.notify != nil {
+		notify = b.notify
+		b.notify = nil
+	}
+
+	// Check to see if we are full.
+	if b.full {
+		// If so, grow AddrPacket buffer by one.
+		b.packets = append(b.packets[:b.write], append(make([]AddrPacket, 1), b.packets[b.write:]...)...)
```

Comment:
In the event that we receive a flood of writes without reads, this becomes quite inefficient. Compare to the [grow logic](https://github.com/pion/transport/blob/6890c795c807a617c054149eee40a69d7fdfbfdb/packetio/buffer.go#L82) of the `packetio.Buffer`, where we are doubling the size of the underlying byte slice until we reach a cutoff point (at which point we grow at a slower rate).

We could opt to use a similar strategy here, though we still perform a large number of allocations due to the need to allocate each `AddrPacket`'s underlying byte slice. I have been able to bring this down to being equivalent (or less than) the number of allocations in the `packetio.Buffer` implementation by instead using an array with size equivalent to the UDP `PacketConn` MTU, which we know we will never exceed here. However, it does mean that we could be holding onto much more memory than needed. For example, if our buffer grows to size 100, the payload on all of those packets could each just be a few bytes, but the total allocated memory for payloads would be `100*MTU`.

`AddrPacket` looks like this with the change:
```
type AddrPacket struct {
	addr net.Addr
	data [8192]byte
	size int
}
```

Reply:
Another option here would be to use a byte slice on `AddrPacket` but use a capped growing strategy for it as well.

Reply:
Chatted with @edaniels -- decision for now is to use a `bytes.Buffer` on `AddrPacket` for now. The `PacketBuffer` handles growing the ring buffer, but allocating the `data` for each packet is deferred to the `bytes.Buffer`. Will revisit opportunities to improve performance in the future.